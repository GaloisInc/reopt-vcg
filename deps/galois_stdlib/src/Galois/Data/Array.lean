namespace Array

section
universe u
variables {α : Type u} [HasLess α] [DecidableRel (HasLess.Less : α → α → Prop)]

partial def ltAux (xs ys : Array α) : Nat → Bool
| n =>
  if hxs : n < xs.size then
    if hys : n < ys.size then
      let xn := xs.get ⟨n, hxs⟩;
      let yn := ys.get ⟨n, hys⟩;
      if  xn < yn then true
      else if yn < xn then false
      else ltAux (n+1)
    else false
  else if xs.size < ys.size then
    true
  else
    false

partial def lt (xs ys : Array α) : Bool := ltAux xs ys 0

end

end Array
