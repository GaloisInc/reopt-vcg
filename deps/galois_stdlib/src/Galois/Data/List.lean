
import Galois.Init.Order

namespace List

universes u v

def qsort {α : Type} [Inhabited α] (as : List α) (lt : α → α → Bool) : List α :=
  let arr := as.toArray;
  (arr.qsort lt 0 (arr.size - 1)).toList


def joinMap {α : Type u} {β : Type v} (f : α → List β) : List α → List β
  | []      => []
  | a :: as => (f a) ++ joinMap f as


-- Powerset of a single list.
def powerset {α : Type u} : List α → List (List α)
| [] => [[]]
| x::xs =>
  joinMap (λ l => [l, x::l]) (powerset xs)

-- Cartesian product of two lists.
def product {α : Type u} {β : Type v} : List α → List β → List (α × β)
| [], _ => []
| a::as, bs => (bs.map (λ b => (a,b)))++(product as bs)

-- List.power xs ys computes ys^xs, or the set of sequences of elements of ys indexed by elements of xs.
def power {α β} : List α → List β → List (List (α × β))
| [] , vs => []
| k::ks, vs => (power ks vs).joinMap (λ m => vs.map (λ v => (k, v)::m))

inductive In {α : Type u} (x : α) : List α → Prop
| first : ∀ (l : List α), In x (x :: l)
| rest  : ∀ (y : α) {xs : List α}, In x xs → In x (y :: xs)

inductive Forall {α : Type u} (p : α → Prop) : List α → Prop
| nil  : Forall p []
| cons : ∀ {x : α} {xs : List α}, p x → Forall p xs → Forall p (x :: xs)

namespace Forall

def map {α : Type u} {p q : α → Prop} (f : ∀ {x : α}, p x → q x) : ∀ {l : List α}, Forall p l → Forall q l
| [], Forall.nil => Forall.nil
| (_::_), Forall.cons xH xsH => Forall.cons (f xH) (map f xsH)

end Forall




inductive SortedMap {α : Type u} {β : Type v} [LessOrder α] : List (α × β) → Prop
  | nil : SortedMap []
  | cons : ∀ (k : α) (v : β) (l : List (α × β)),
           SortedMap l →
           l.Forall (λ (kv : (α × β)) => k < kv.fst) →
           -- N.B., HasLessOrder combined with this Forall implies
           -- no duplicate keys.
           SortedMap ((k,v)::l)


namespace SortedMap
variables {α : Type u} {β : Type v}

protected def insert [DecidableLessOrder α] (k : α) (v : β) : List (α × β) →  List (α × β)
| [] => [(k,v)]
| (k0, v0)::rst =>
  if k < k0
  then (k,v)::(k0, v0)::rst
  else if k0 < k
  then (k0, v0)::(SortedMap.insert k v rst)
  else (k,v)::rst

-- FIXME prove
axiom insert.wellFormed [DecidableLessOrder α] (k : α) (v : β) {l : List (α × β)} :
List.SortedMap l →
List.SortedMap (SortedMap.insert k v l)

-- FIXME prove
axiom insert.stillNotIn [DecidableLessOrder α] (k : α) {v v' : β} (pf : v ≠ v') {l : List (α × β)} :
l.Forall (λ (kv : α × β) => kv.snd ≠ v') →
(SortedMap.insert k v l).Forall (λ (kv : α × β) => kv.snd ≠ v')


protected def erase [DecidableLessOrder α] (k : α) : List (α × β) →  List (α × β)
| [] => []
| l@((k0, v0)::rst) =>
  if k = k0
  then rst
  else if k0 > k
  then l
  else (k0,v0)::(SortedMap.erase k rst)

-- FIXME prove
axiom erase.wellFormed [DecidableLessOrder α] (k : α) {l : List (α × β)} :
List.SortedMap l →
List.SortedMap (SortedMap.erase k l)

-- FIXME prove
axiom erase.stillNotIn [DecidableLessOrder α] (k : α) {v : β} {l : List (α × β)} :
l.Forall (λ (kv : α × β) => kv.snd ≠ v) →
(SortedMap.erase k l).Forall (λ (kv : α × β) => kv.snd ≠ v)


end SortedMap


end List
