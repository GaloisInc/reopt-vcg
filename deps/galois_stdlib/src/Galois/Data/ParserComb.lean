-- Parsing combinators

import Init.System.IO
import Init.Control.State
import Init.Control.Id
import Init.Control.Option

-- Single-valued, backtracking on failure
structure Parser (tok : Type) (a : Type) :=
  ( run : StateT (List tok) (OptionT Id) a )

namespace Parser

variables {tok a b : Type}

instance : HasOrelse (Parser tok a) := 
  { orelse := fun x y => Parser.mk (x.run <|> y.run) }

instance : Monad (Parser tok) :=
  { bind := fun _ _ x f => Parser.mk (x.run >>= fun v => (f v).run)
  , pure := fun _ x => Parser.mk (pure x)
  }

instance : MonadStateOf (List tok) (Parser tok) :=
  { set := fun s => Parser.mk (set s)
  , get := Parser.mk get
  , modifyGet := fun _ f => Parser.mk (modifyGet f)
  }

instance : MonadExcept Unit (Parser tok) :=
  { throw := fun _ e => Parser.mk (throw e)
  , catch := fun _ m f => Parser.mk (catch (m.run) (fun x => (f x).run))
  }

protected
partial def many0 (p : Parser tok a) : Unit -> Parser tok (List a)
  | _ => (do v <- p; vs <- many0 (); pure (v :: vs)) <|> pure []

def optional {a : Type} (f : tok -> Option a) : Parser tok a :=
  do tks <- get;
     match tks with     
       | [] => throw ()
       | (t :: tks') => 
         match f t with 
           | none => throw ()
           | some r => (do set tks'; pure r)

def token (f : tok -> Bool) : Parser tok tok :=
  optional (fun t => if f t then some t else none)

def exact [DecidableEq tok] (t : tok) : Parser tok tok :=
  token (fun x => x = t)

def many (p : Parser tok a) : Parser tok (List a) := Parser.many0 p ()

def many1 (p : Parser tok a) : Parser tok (List a) := 
  do v <- p; 
     vs <- many p;
     pure (v :: vs)

def sepBy (s : Parser tok b) (p : Parser tok a) : Parser tok (List a) := 
  (do rs <- many (do x <- p; discard s; pure x); -- p <* s
      -- rs <- pure [];
      r  <- p;
      pure (List.append rs [r])) <|> pure []

def first (ps : List (Parser tok a)) : Parser tok a :=
  List.foldl (fun p m => p <|> m) (throw ()) ps

def runParser (p : Parser tok a) (tks : List tok) : Option a :=
  (p.run.run' tks).run.run

end Parser

@[reducible]
def CharParser := Parser Char

namespace CharParser
open Parser

def digitP : CharParser Nat :=
  optional (fun c => if c.isDigit then some (c.toNat - '0'.toNat) else none)

def natP : CharParser Nat :=
  do ds <- many1 digitP;
     pure (ds.foldl (fun acc d => acc * 10 + d) 0)

def intP : CharParser Int :=
  (do discard $ exact '-';
      n <- natP;
      pure (Int.negOfNat n)) <|> 
      (Int.ofNat <$> natP)

def stringP (f : Char -> Bool) : CharParser String := List.asString <$> many (token f)
def string1P (f : Char -> Bool) : CharParser String := List.asString <$> many1 (token f)

def nonWSP : CharParser String := 
  string1P (fun c => not (Char.isWhitespace c))

def whitespaceP : CharParser Unit := 
do _ <-token Char.isWhitespace;
   pure ()

def parseFile {a : Type} {m : Type â†’ Type} [Monad m] [MonadIO m] (p : CharParser a) (fname : String)
  : m (Option a) := 
  do s <- IO.FS.readFile fname;
     pure (runParser p s.toList)
  

end CharParser
