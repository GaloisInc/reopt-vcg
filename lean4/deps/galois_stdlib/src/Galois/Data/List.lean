namespace List

section
universes u v
variables {α : Type u} {β : Type v}

def joinMap (f : α → List β) : List α → List β
| []      => []
| a :: as => (f a) ++ joinMap as

end

def lexLt {α} (lt : α → α → Bool) : List α → List α → Bool
| [], [] => false
| [], y::ys => true
| x::xs, [] => false
| x::xs, y::ys =>
  if lt x y then true
  else if lt y x then false
  else lexLt xs ys

end List

namespace SortedAList
universes u v
variables {α : Type u} {β : Type v} (lt : α → α → Bool)


protected def insert (k : α) (v : β) : List (α × β) →  List (α × β)
| [] => [(k,v)]
| (k0, v0)::rst =>
  if lt k k0
  then (k,v)::(k0, v0)::rst
  else if lt k0 k
  then (k0, v0)::(insert rst)
  else (k,v)::rst

end SortedAList

