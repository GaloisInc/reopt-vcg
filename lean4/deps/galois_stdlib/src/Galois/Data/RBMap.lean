
namespace RBMap
universes u v w
variables {α : Type u} {β : Type v}

/-- Builds an RBMap from a list with String keys. --/
def ltMap [HasLess α] [∀ (a b : α), Decidable (a < b)] (entries: List (α × β))
 : RBMap α β (λ (x y:α)=> x<y) :=
RBMap.fromList entries (λ (x y:α)=> x<y)


section
variable {lt : α → α → Bool}

@[specialize] def keys : RBMap α β lt → List α
| ⟨t, _⟩ => t.revFold (fun ps k _v => k::ps) []

@[specialize] def values : RBMap α β lt → List β
| ⟨t, _⟩ => t.revFold (fun ps _k v => v::ps) []
end

end RBMap
