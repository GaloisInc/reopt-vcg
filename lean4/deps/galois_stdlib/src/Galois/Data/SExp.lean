
-- Here we defined data and functions for
-- well-formed s-expressions (i.e., s-expressions
-- which exclude improper lists by construction).
namespace WellFormedSExp


section
universe u

inductive SExp (α : Type u)
| atom : α → SExp
| list : List SExp → SExp


-- FIXME consider using wf rec here when it's enabled in lean4
partial def SExp.toString {α : Type u} [HasToString α] : SExp α → String
| SExp.atom a => toString a
| SExp.list ss => "(" ++ (String.join $ (ss.map SExp.toString).intersperse " ")++ ")"


namespace SExp
open SExp

instance {α : Type u} [HasToString α] : HasToString (SExp α) := ⟨SExp.toString⟩

inductive Tag
| atom
| nonAtom

inductive ParseState (α : Type u) : Tag → Type u
| init : ParseState Tag.nonAtom
| inAtom : Char → List Char → ParseState Tag.nonAtom → ParseState Tag.atom
| inList : List (SExp α) → ParseState Tag.nonAtom → ParseState Tag.nonAtom

open ParseState

private def parseFail {α : Type u} (specific: String) : Except String (SExp α × String) :=
Except.error $ "Failed to parse s-expression: " ++ specific ++ "."

private def revStr (c:Char) (cs:List Char) : String := (c::cs).reverse.asString


def readAux {α : Type u} (parseAtom : String → Except String α) :
List Char →
Sigma (ParseState α) → 
Except String (SExp α × String)
| [], state => 
  match state with
  | ⟨_, init⟩ => parseFail $ "input was empty"
  | ⟨_, inAtom x xs aPrev⟩ =>
    match aPrev with
    | init => do
      a ← parseAtom $ revStr x xs;
      pure (atom a, "")
    | inList _ _ => parseFail "missing closing parenthesis"
  | ⟨_, inList _ _⟩ => parseFail "missing closing parenthesis"
| (c::cs), state =>
  match state with
  | ⟨_, init⟩ =>
    if c.isWhitespace then readAux cs ⟨Tag.nonAtom, init⟩
    else if c == '(' then readAux cs $ ⟨Tag.nonAtom, inList [] init⟩
    else if c == ')' then parseFail "unexpected closing parenthesis"
    else readAux cs $ ⟨Tag.atom, inAtom c [] init⟩
  | ⟨_, inAtom x xs prev⟩ =>
    if c.isWhitespace then 
      match prev with
      | init => do
        a ← parseAtom $ revStr x xs;
        pure (atom a, "")
      | inList ys ysPrev => do
        a ← parseAtom $ revStr x xs;
        readAux cs $ ⟨Tag.nonAtom, inList (atom a::ys) ysPrev⟩
    else if c == '(' then do
        a ← parseAtom $ revStr x xs;
        match prev with
        | init => pure (atom a, (c::cs).asString)
        | inList zs zsPrev => readAux cs $ ⟨Tag.nonAtom, inList [] (inList (atom a::zs) zsPrev)⟩
    else if c == ')' then do
        a ← (parseAtom $ revStr x xs).map atom;
        match prev with
        | init => pure (a, (c::cs).asString)
        | inList ys ysPrev =>
          let l := list (a::ys).reverse;
          match ysPrev with
          | init => pure (l, cs.asString)
          | inList zs zsPrev => readAux cs $ ⟨Tag.nonAtom, inList (l::zs) zsPrev⟩
    else readAux cs $ ⟨Tag.atom, inAtom c (x::xs) prev⟩
  | ⟨_, inList xs xsPrev⟩ =>
    if c.isWhitespace then readAux cs $ state
    else if c == '(' then readAux cs $ ⟨Tag.nonAtom, inList [] $ inList xs xsPrev⟩
    else if c == ')' then
      match xsPrev with
      | init => pure (list xs.reverse, cs.asString)
      | inList zs zsPrev => readAux cs $ ⟨Tag.nonAtom, inList ((list xs.reverse)::zs) zsPrev⟩
    else readAux cs $ ⟨Tag.atom, inAtom c [] $ inList xs xsPrev⟩


/-- Read the first s-expression in the string, using the given atom parser
    to convert atom strings into the parameter type `α`. The result
    is either an error message or the parsed s-expression and the
    remainder of the string. --/
def read 
{α : Type u}
(parseAtom : String → Except String α)
(str: String) 
: Except String (SExp α × String) :=
readAux parseAtom str.data ⟨Tag.nonAtom, init⟩

/-- Like `read` but does not return the remainder of the input string. --/
def readSExp
{α : Type u}
(parseAtom : String → Except String α)
(str: String)
: Except String (SExp α) := do
(s, _) ← read parseAtom str;
pure s

/-- Like `readSExp` but reads all s-expressions from the input string and returns
    them in a list. --/
def readSExps
{α : Type u}
(parseAtom : String → Except String α)
(str: String)
: Except String (List (SExp α)) := do
(ss, _) ← readAux parseAtom (str.data ++ [')']) ⟨Tag.nonAtom, inList [] init⟩;
match ss with
| atom a => pure [ss]
 -- ^ not technically possible given how the parser works but that's not locally obvious
| list l => pure l


end SExp

end

end WellFormedSExp
