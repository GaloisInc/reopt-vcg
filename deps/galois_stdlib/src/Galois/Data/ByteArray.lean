import Galois.Data.Array


namespace ByteArray

def lt (xs ys : ByteArray) : Bool := Array.lt xs.data ys.data


private def appendCharAsBytes (c:Char) (bs : List UInt8) : List UInt8 :=
  let b1  : UInt8 := UInt8.ofNat $ (c.val.land 0x000000FF).toNat;
  let b2  : UInt8 := UInt8.ofNat $ (c.val.land 0x0000FF00).toNat;
  let b3  : UInt8 := UInt8.ofNat $ (c.val.land 0x00FF0000).toNat;
  let b4  : UInt8 := UInt8.ofNat $ (c.val.land 0xFF000000).toNat;
  b1::b2::b3::b4::bs

def fromString (s:String) : ByteArray := 
   (s.data.foldr appendCharAsBytes []).toByteArray

end ByteArray
