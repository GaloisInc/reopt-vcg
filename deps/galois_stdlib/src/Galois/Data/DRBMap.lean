
import Std.Data.RBMap

namespace Sigma
universes u v
variables {α : Type u} {β : α → Type v}

instance [Inhabited α] [forall (a : α), Inhabited (β a)] : Inhabited (Sigma β) :=
{default := let a : α := arbitrary; ⟨a, (arbitrary : (β a)) ⟩}

end Sigma



namespace Std
universes u v w w'

namespace RBNode
variables {α : Type u} {β : α → Type v} (lt : α → α → Bool)

@[specialize] def findEq [DecidableEq α] : RBNode α β → ∀ (k : α), Option (β k)
| RBNode.leaf,             x => none
| RBNode.node _ a ky vy b, x =>
  if hEq : ky = x
  then some $ cast (congrArg β hEq) vy
  else if lt x ky then findEq a x
  else findEq b x

end RBNode

open Std.RBNode

def DRBMap (α : Type u) (β : α → Type v) (lt : α → α → Bool) : Type (max u v) :=
{t : RBNode α β // t.WellFormed lt }


@[inline] def mkDRBMap (α : Type u) (β : α → Type v) (lt : α → α → Bool) : DRBMap α β lt :=
⟨leaf, WellFormed.leafWff⟩

@[inline] def DRBMap.empty {α : Type u} {β : α → Type v} {lt : α → α → Bool} : DRBMap α β lt :=
mkDRBMap _ _ _

-- instance DRBMap.HasEmptyc (α : Type u) (β : α → Type v) (lt : α → α → Bool) : HasEmptyc (DRBMap α β lt) :=
-- ⟨DRBMap.empty⟩

namespace DRBMap
variables {α : Type u} {β : α → Type v} {σ : Type w} {lt : α → α → Bool}


def depth (f : Nat → Nat → Nat) (t : DRBMap α β lt) : Nat :=
t.val.depth f

@[inline] def fold (f : σ → forall (a : α), (β a) → σ) : σ → DRBMap α β lt → σ
| b, ⟨t, _⟩ => t.fold f b

@[inline] def revFold (f : σ → forall (a : α), (β a) → σ) : σ → DRBMap α β lt → σ
| b, ⟨t, _⟩ => t.revFold f b

@[inline] def foldM {m : Type w → Type w'} [Monad m] (f : σ → forall (a : α), (β a) → m σ) : σ → DRBMap α β lt → m σ
| b, ⟨t, _⟩ => t.foldM f b

@[inline] def forM {m : Type w → Type w'} [Monad m] (f : forall (a : α), (β a) → m PUnit) (t : DRBMap α β lt) : m PUnit :=
t.foldM (fun _ k v => f k v) ⟨⟩

@[inline] def isEmpty : DRBMap α β lt → Bool
| ⟨leaf, _⟩ => true
| _         => false

@[specialize] def toList : DRBMap α β lt → List (Sigma β)
| ⟨t, _⟩ => t.revFold (fun ps k v => ⟨k, v⟩::ps) []

@[inline] protected def min : DRBMap α β lt → Option (Sigma β)
| ⟨t, _⟩ =>
  match t.min with
  | some ⟨k, v⟩ => some ⟨k, v⟩
  | none        => none

@[inline] protected def max : DRBMap α β lt → Option (Sigma β)
| ⟨t, _⟩ =>
  match t.max with
  | some ⟨k, v⟩ => some ⟨k, v⟩
  | none        => none

instance [Repr α] [forall (a : α), Repr (β a)] : Repr (DRBMap α β lt) :=
⟨fun t => "rbmapOf " ++ repr t.toList⟩

@[inline] def insert : DRBMap α β lt → forall (a : α), (β a) → DRBMap α β lt
| ⟨t, w⟩, k, v => ⟨t.insert lt k v, WellFormed.insertWff w rfl⟩

@[inline] def erase : DRBMap α β lt → α → DRBMap α β lt
| ⟨t, w⟩, k => ⟨t.erase lt k, WellFormed.eraseWff w rfl⟩

@[specialize] def ofList : List (Sigma β) → DRBMap α β lt
| []        => mkDRBMap _ _ _
| ⟨k,v⟩::xs => (ofList xs).insert k v

@[inline] def find? : DRBMap α β lt → α → Option (Sigma β)
| ⟨t, _⟩, x => t.findCore lt x

@[inline] def findEq? [DecidableEq α] : DRBMap α β lt → forall (a : α), Option (β a)
| ⟨t, _⟩, x => t.findEq lt x

@[inline] def findD (t : DRBMap α β lt) (k : α) (v₀ : Sigma β) : Sigma β :=
(t.find? k).getD v₀

@[inline] def findEqD [DecidableEq α] (t : DRBMap α β lt) (k : α) (v₀ : β k) : β k :=
(t.findEq? k).getD v₀


/-- (lowerBound k) retrieves the kv pair of the largest key smaller than or equal to `k`,
    if it exists. -/
@[inline] def lowerBound : DRBMap α β lt → α → Option (Sigma β)
| ⟨t, _⟩, x => t.lowerBound lt x none

@[inline] def contains (t : DRBMap α β lt) (a : α) : Bool :=
(t.find? a).isSome

@[inline] def fromList (l : List (Sigma β)) (lt : α → α → Bool) : DRBMap α β lt :=
l.foldl (fun r p => r.insert p.1 p.2) (mkDRBMap α β lt)

@[inline] def all : DRBMap α β lt → (forall (a : α), β a → Bool) → Bool
| ⟨t, _⟩, p => t.all p

@[inline] def any : DRBMap α β lt → (forall (a : α), β a → Bool) → Bool
| ⟨t, _⟩, p => t.any p

def size (m : DRBMap α β lt) : Nat :=
m.fold (fun sz _ _ => sz+1) 0

def maxDepth (t : DRBMap α β lt) : Nat :=
t.val.depth Nat.max


@[inline] def min! [Inhabited α] [forall (a : α), Inhabited (β a)] (t : DRBMap α β lt) : Sigma β :=
match t.min with
| some p => p
| none   => panic! "map is empty"

@[inline] def max! [Inhabited α] [forall (a : α), Inhabited (β a)] (t : DRBMap α β lt) : Sigma β :=
match t.max with
| some p => p
| none   => panic! "map is empty"

@[inline] def find! [Inhabited α] [forall (a : α), Inhabited (β a)] (t : DRBMap α β lt) (k : α) : Sigma β :=
match t.find? k with
| some b => b
| none   => panic! "key is not in the map"

@[inline] def findEq! [DecidableEq α] [Inhabited α] [forall (a : α), Inhabited (β a)] (t : DRBMap α β lt) (k : α) : β k :=
match t.findEq? k with
| some b => b
| none   => panic! "key is not in the map"


end DRBMap

def drbmapOf {α : Type u} {β : α → Type v} (l : List (Sigma β)) (lt : α → α → Bool) : DRBMap α β lt :=
DRBMap.fromList l lt

end Std
