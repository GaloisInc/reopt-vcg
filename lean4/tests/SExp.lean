
import Galois.Data.SExp

namespace Test

namespace SExp

structure Atom :=
(val:String)

def Atom.toString (a:Atom) : String := a.val

instance Atom.HasToString : HasToString Atom := ⟨Atom.toString⟩

def parseAtom (input:String) : Except String Atom := pure $ Atom.mk input

abbrev SExpr := WellFormedSExp.SExp Atom

def parsePass (input : String) : IO String :=
match WellFormedSExp.SExp.readSExp parseAtom input with
| Except.error errMsg => 
  pure $ "unexpected fail on `" ++ input ++ "` with error message `"++errMsg++"`"
| Except.ok sexp => pure $ "pass " ++ sexp.toString

def parseFail (input : String) : IO String :=
match WellFormedSExp.SExp.readSExp parseAtom input with
| Except.error _errMsg => pure $ "fail " ++ input
| Except.ok sexp => pure $ "unexpected pass on `" ++ input ++ "` which parsed as `"++sexp.toString++"`."


def test : IO UInt32 := do
-- Tests that should pass
parsePass "42" >>= IO.println;
parsePass "x" >>= IO.println;
parsePass "()" >>= IO.println;
parsePass "(())" >>= IO.println;
parsePass "(x)" >>= IO.println;
parsePass "( x )" >>= IO.println;
parsePass "(x y)" >>= IO.println;
parsePass "(x(y(z)))" >>= IO.println;
parsePass "(x (y (z)))" >>= IO.println;
-- Tests that should fail
parseFail "(" >>= IO.println;
parseFail "(x" >>= IO.println;
parseFail "(() x" >>= IO.println;
pure 0

end SExp
end Test


