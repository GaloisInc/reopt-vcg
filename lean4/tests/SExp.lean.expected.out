pass 42
pass x
pass ()
pass (())
pass (x)
pass (x)
pass (x y)
pass (x (y (z)))
pass (x (y (z)))
fail (
fail (x
fail (() x
